### PR TITLE
[export] Symint support (nonstrict, Dim.DYNAMIC)

### LINKED ISSUE #113682: [export] Support capturing of non-tensor inputs
### 🚀 The feature, motivation and pitch

Currently, `torch.export.export` would burn-in non-tensor inputs:

Example:
```python
import torch

def logsoftmax(x, dim):
    return torch.nn.LogSoftmax(dim)(x)

exp = torch.export.export(logsoftmax, ( torch.randn(10, 10), 1))
print(exp.graph_module.code)
```
yields 

```
def forward(self, arg0_1, arg1_1):
    _log_softmax = torch.ops.aten._log_softmax.default(arg0_1, 1, False);  arg0_1 = None
    return (_log_softmax,)
```

Here, arg1_1 become dangling and the value `1` is hardcoded. So when this function is called with a different value, it would produce wrong result.

Desired output would be
```
def forward(self, arg0_1, arg1_1):
    _log_softmax = torch.ops.aten._log_softmax.default(arg0_1, arg1_1, False);  arg0_1 = None
    return (_log_softmax,)
```

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by angelayi**
Fixes https://github.com/pytorch/pytorch/issues/113682 only in the non-strict export case. Also we only support Dim.DYNAMIC/AUTO, not named-Dims

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/_export/non_strict_utils.py [hunk 1]
```diff
@@ -95,6 +97,28 @@ def fakify(
...
+                    else None
+                )
+                mode.shape_env.tracked_fakes.append(  # type: ignore[union-attr]
+                    TrackedFake(symint, source, context)
+                )
+                return symint
+            else:
+                return t.val
+        else:
+            return t.val
```
Comment (pianpwk) [inline torch/_export/non_strict_utils.py:120]: nit: merge these 2 elses?

File: torch/_export/non_strict_utils.py [hunk 2]
```diff
@@ -391,61 +456,79 @@ def is_int(x: object) -> bool:
...
-        if input_index < num_lifted_inputs or node.op != "placeholder":
-            continue
-        if _is_constant_argument(node.meta["val"]) or isinstance(
-            node.meta["val"], CustomObjArgument
+        meta_val = node.meta.get("val")
+
+        if (
+            input_index < num_lifted_inputs
+            or node.op != "placeholder"
+            or meta_val is None
```
Comment (pianpwk) [inline torch/_export/non_strict_utils.py:467]: hm why the change here?

File: torch/_export/non_strict_utils.py [hunk 3]
```diff
@@ -391,61 +456,79 @@ def is_int(x: object) -> bool:
...
-            # we might want to do this even if model tracing inferred a static dimension.
-            if isinstance(dim, _DimHint):
-                trace_vr = (
-                    range_constraints[d.node.expr]
-                    if not is_int(d)
-                    else ValueRanges(int(d), int(d))
+        elif isinstance(meta_val, torch.SymInt):
+            if shape_spec is not None and isinstance(shape_spec, _DimHint):
+                range_constraints[meta_val.node.expr] &= shape_env.bound_sympy(
+                    meta_val.node._expr
```
Comment (pianpwk) [inline torch/_export/non_strict_utils.py:485]: hmm is the _constrain_user_specified_dimhint_range call not enough here?

File: torch/_export/utils.py [hunk 4]
```diff
@@ -282,33 +282,109 @@ def _rename_without_collisions(
...
+    if args_kwargs_key_path.idx == 0:
+        return f"*args{keystr(key_path[1:])}"
+    else:
+        kwarg_key = key_path[1]
+        assert isinstance(kwarg_key, MappingKey)
+        name = str(kwarg_key)[1:-1]  # get rid of the enclosed []
+        return f"{name}{keystr(key_path[2:])}"
+
+
+def _check_symint(
```
Comment (pianpwk) [inline torch/_export/utils.py:303]: looks like just a pure refactor?

File: torch/export/dynamic_shapes.py [hunk 5]
```diff
@@ -1022,10 +1041,27 @@ def root_value():
...
+        elif isinstance(dim, _DimHint):
+            if dim.type == _DimHintType.AUTO:
+                torch._dynamo.maybe_mark_dynamic(tensor, idx)
+            elif dim.type == _DimHintType.STATIC:
+                torch._dynamo.mark_static(tensor, idx)
+            elif dim.type == _DimHintType.DYNAMIC:
+                torch._dynamo.mark_dynamic(tensor, idx)
+            constraints.append(_RelaxedConstraint(id(tensor), idx))
+        elif dim is None:
+            torch._dynamo.mark_static(tensor, idx)
```
Comment (pianpwk) [inline torch/export/dynamic_shapes.py:1062]: thanks, this is nice

**Review by pianpwk — APPROVED** [review]

File: torch/_export/non_strict_utils.py (hunk 2 above)
Comment (angelayi) [reply torch/_export/non_strict_utils.py:467]: I just added the `meta_val is None` check, just in case haha

File: torch/_export/utils.py (hunk 4 above)
Comment (angelayi) [reply torch/_export/utils.py:303]: I took out this section from `_check_input_constraints_for_graph` and I call it in both the tensor shape part (line 421) and the symint part (line 430)
